### PR TITLE
mgr/cephadm: Allow idmap overrides in nfs-ganesha configuration

### PR DESCRIPTION
**Pull request by asm0deuz**
Fixes: https://tracker.ceph.com/issues/64577

### CODE REVIEW

**Review by adk3798 — COMMENTED** [review]
It seems like all the pieces are here when I look this over. @asm0deuz have you been able to test this out?

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 1]
```diff
@@ -115,12 +119,28 @@ def get_ganesha_conf() -> str:
...
                 context["haproxy_hosts"] = self._haproxy_hosts()
                 logger.debug("selected haproxy_hosts: %r", context["haproxy_hosts"])
             return self.mgr.template.render('services/nfs/ganesha.conf.j2', context)
 
+        # generate the idmap config
+        def get_idmap_conf() -> str:
+            idmap_conf = spec.idmap_conf
+            cp = ConfigParser()
+            out = StringIO()
+            for sections, keys in idmap_conf:
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/nfs.py:134]: mypy seems unhappy with this
```
cephadm/services/nfs.py: note: In function "generate_config":
cephadm/services/nfs.py:134: error: Unpacking a string is disallowed
cephadm/services/nfs.py: note: In function "get_idmap_conf":
cephadm/services/nfs.py:134: error: Unpacking a string is disallowed
cephadm/services/nfs.py: note: In function "generate_config":
cephadm/services/nfs.py:134: error: Item "None" of "Optional[Dict[str, Dict[str, str]]]" has no attribute "__iter__" (not iterable)
cephadm/services/nfs.py: note: In function "get_idmap_conf":
cephadm/services/nfs.py:134: error: Item "None" of "Optional[Dict[str, Dict[str, str]]]" has no attribute "__iter__" (not iterable)
```
I'd guess you need to wrap part of this function in `if idmap_conf is not None`

**Comment by asm0deuz** [comment]
@adk3798 I'll test that tomorrow morning and let you know the outcome.

File: src/pybind/mgr/cephadm/services/nfs.py (hunk 1 above)
Comment (asm0deuz) [reply src/pybind/mgr/cephadm/services/nfs.py]: As I replaced the iterations by .read_dict() it should be ok now but I'll check the outcome of mypy

**Comment by asm0deuz** [comment]
@adk3798  Looks like it works as expected:

nfs spec file used:

```
service_type: nfs
service_id: mynfs
placement:
  hosts:
    - node-1
spec:
  port: 12345
idmap_conf:
  general:
    Local-Realms: nowhere.org
  mapping:
    Nobody-User: nobody
    Nobody-Group: nobody
```


```
[ceph: root@node-1 /]# ceph orch ls
NAME                       PORTS    RUNNING  REFRESHED  AGE  PLACEMENT  
crash                                   1/1  2m ago     8m   *          
mgr                                     2/2  2m ago     8m   count:2    
mon                                     1/5  2m ago     8m   count:5    
nfs.mynfs                  ?:12345      1/1  2m ago     3m   node-1     
osd.all-available-devices                 4  2m ago     7m   *          
```

```
[root@node-1 ganesha]# cat idmap.conf 
# This file is generated by cephadm.
[general]
local-realms = nowhere.org

[mapping]
nobody-user = nobody
nobody-group = nobody
```

```
[root@node-1 ganesha]# cat /etc/ganesha/ganesha.conf 
# This file is generated by cephadm.
NFS_CORE_PARAM {
        Enable_NLM = false;
        Enable_RQUOTA = false;
        Protocols = 4;
        NFS_Port = 12345;
}

NFSv4 {
        Delegations = false;
        RecoveryBackend = 'rados_cluster';
        Minor_Versions = 1, 2;
        IdmapConf = "/etc/ganesha/idmap.conf";
}

RADOS_KV {
        UserId = "nfs.mynfs.0.1.node-1.fgbkwz";
        nodeid = "nfs.mynfs.0";
        pool = ".nfs";
        namespace = "mynfs";
}

RADOS_URLS {
        UserId = "nfs.mynfs.0.1.node-1.fgbkwz";
        watch_url = "rados://.nfs/mynfs/conf-nfs.mynfs";
}

RGW {
        cluster = "ceph";
        name = "client.nfs.mynfs.0.1.node-1.fgbkwz-rgw";
}
```

**Comment by adk3798** [comment]
> @adk3798 Looks like it works as expected:
> 
> nfs spec file used:
> 
> ```
> service_type: nfs
> service_id: mynfs
> placement:
>   hosts:
>     - node-1
> spec:
>   port: 12345
> idmap_conf:
>   general:
>     Local-Realms: nowhere.org
>   mapping:
>     Nobody-User: nobody
>     Nobody-Group: nobody
> ```
> 
> ```
> [ceph: root@node-1 /]# ceph orch ls
> NAME                       PORTS    RUNNING  REFRESHED  AGE  PLACEMENT  
> crash                                   1/1  2m ago     8m   *          
> mgr                                     2/2  2m ago     8m   count:2    
> mon                                     1/5  2m ago     8m   count:5    
> nfs.mynfs                  ?:12345      1/1  2m ago     3m   node-1     
> osd.all-available-devices                 4  2m ago     7m   *          
> ```
> 
> ```
> [root@node-1 ganesha]# cat idmap.conf 
> [general]
> local-realms = nowhere.org
> 
> [mapping]
> nobody-user = nobody
> nobody-group = nobody
> ```

what directory was this run from? If it was `/var/lib/ceph/<fsid>/<nfs-daemon-name>/etc/ganesha/` then it seems like we're good. Last thing to check would just be that it's mounted inside the container properly.

**Comment by asm0deuz** [comment]
@adk3798 

> what directory was this run from? If it was `/var/lib/ceph/<fsid>/<nfs-daemon-name>/etc/ganesha/` then it seems like we're good. Last thing to check would just be that it's mounted inside the container properly.

First and second output comes from the cephadm shell while the idmap.conf output comes from within the nfs container.
Just fixed one remaining failing unitest.

**Review by adk3798 — APPROVED** [review]
code seems good to me

**Comment by adk3798** [comment]
jenkins test make check

**Comment by asm0deuz** [comment]
jenkins test make check

**Comment by asm0deuz** [comment]
https://pulpito.ceph.com/tonay-2024-03-04_15:07:38-orch:cephadm-wip-asm0deuz-testing-2024-03-04-1211-distro-default-smithi/

**Comment by adk3798** [comment]
> https://pulpito.ceph.com/tonay-2024-03-04_15:07:38-orch:cephadm-wip-asm0deuz-testing-2024-03-04-1211-distro-default-smithi/

lots of unlucky infra/teuthology failures and hitting a bunch of the failures from scraping the logs at the end (still working on the ignorelist for the cephadm suite), but nothing I can see that was caused by your PR, and the only nfs tests failing are doing so for `Failed to reconnect to...` reasons, which is unrelated to this PR. So, I don't think this PR is causing any regressions.

**Pull request merged**